src/Backups/BackupIO_Disk.cpp

```diff
@@ -8,138 +8,122 @@
 namespace DB
 {
 
-namespace ErrorCodes
-{
-    extern const int LOGICAL_ERROR;
-}
-
-BackupReaderDisk::BackupReaderDisk(const DiskPtr & disk_, const String & path_)
-    : disk(disk_), path(path_), log(&Poco::Logger::get("BackupReaderDisk"))
+BackupReaderDisk::BackupReaderDisk(const DiskPtr & disk_, const String & root_path_, const ContextPtr & context_)
+    : BackupReaderDefault(&Poco::Logger::get("BackupReaderDisk"), context_)
+    , disk(disk_)
+    , root_path(root_path_)
+    , data_source_description(disk->getDataSourceDescription())
 {
 }
 
 BackupReaderDisk::~BackupReaderDisk() = default;
 
 bool BackupReaderDisk::fileExists(const String & file_name)
 {
-    return disk->exists(path / file_name);
+    return disk->exists(root_path / file_name);
 }
 
 UInt64 BackupReaderDisk::getFileSize(const String & file_name)
 {
-    return disk->getFileSize(path / file_name);
+    return disk->getFileSize(root_path / file_name);
 }
 
 std::unique_ptr<SeekableReadBuffer> BackupReaderDisk::readFile(const String & file_name)
 {
-    return disk->readFile(path / file_name);
+    return disk->readFile(root_path / file_name, read_settings);
 }
 
-void BackupReaderDisk::copyFileToDisk(const String & file_name, size_t size, DiskPtr destination_disk, const String & destination_path,
-                                      WriteMode write_mode, const WriteSettings & write_settings)
+void BackupReaderDisk::copyFileToDisk(const String & path_in_backup, size_t file_size, bool encrypted_in_backup,
+                                      DiskPtr destination_disk, const String & destination_path, WriteMode write_mode)
 {
-    if (write_mode == WriteMode::Rewrite)
+    /// Use IDisk::copyFile() as a more optimal way to copy a file if it's possible.
+    bool has_throttling = disk->isRemote() ? static_cast<bool>(read_settings.remote_throttler) : static_cast<bool>(read_settings.local_throttler);
+    if (!has_throttling && (write_mode == WriteMode::Rewrite) && !encrypted_in_backup)
```

Comment:
Do we need to fall back to manual copy if throttling is configured? I think we need to limit the incoming and outgoing traffic only. For server-side copy it would be more natural to put some constraints on the number of in-fly requests.

Reply:
We only use throttling for reading and `IDisk::copyFile()` can't use throttling for reading.